lms/services/canvas_studio.py

```diff
@@ -0,0 +1,221 @@
+from typing import Literal, Optional, TypedDict
+from urllib.parse import urlencode, urlunparse
+
+from marshmallow import EXCLUDE, Schema, fields, post_load
+
+from lms.models.oauth2_token import Service
+from lms.services.aes import AESService
+from lms.services.oauth_http import factory as oauth_http_factory
+from lms.validation._base import RequestsResponseSchema
+
+
+class CanvasStudioCollectionsSchema(RequestsResponseSchema):
+    """Schema for Canvas Studio /collections responses."""
+
+    class CollectionSchema(Schema):
+        class Meta:
+            unknown = EXCLUDE
+
+        id = fields.Integer(required=True)
+        name = fields.Str(required=True)
+        type = fields.Str(required=True)
+        created_at = fields.Str(required=False)
+
+    collections = fields.List(fields.Nested(CollectionSchema), required=True)
+
+    @post_load
+    def post_load(self, data, **_kwargs):
+        return data["collections"]
+
+
+class CanvasStudioCollectionMediaSchema(RequestsResponseSchema):
+    """Schema for Canvas Studio /collections/{id}/media responses."""
+
+    class MediaSchema(Schema):
+        class Meta:
+            unknown = EXCLUDE
+
+        id = fields.Integer(required=True)
+        title = fields.Str(required=True)
+        created_at = fields.Str(required=False)
+
+    media = fields.List(fields.Nested(MediaSchema), required=True)
+
+    @post_load
+    def post_load(self, data, **_kwargs):
+        return data["media"]
+
+
+class APICallInfo(TypedDict):
+    path: str
+    authUrl: str | None
+
+
+class File(TypedDict):
+    """Represents a file or folder in an LMS's file storage."""
+
+    type: Literal["File", "Folder"]
+
+    id: str
+    display_name: str
+    updated_at: str
+
+    contents: Optional[APICallInfo]
+    """API call to use to fetch contents of a folder."""
+
+
+class CanvasStudioService:
+    """
+        Service for authenticating with and making calls to the Canvas Studio API.
+
+    Useful links:
+
+        Authorization guide: https://community.canvaslms.com/t5/Canvas-Studio-Blog/Connecting-Studio-OAuth-via-Postman/ba-p/259739
+        API reference: https://tw.instructuremedia.com/api/public/docs/
+    """
+
+    def __init__(self, request):
+        application_instance = request.lti_user.application_instance
```

Comment:
I'd say prefered way to build the services is to do all of this in the factory and pass this information / services as parameters here.

[Eg. the VS one](https://github.com/hypothesis/lms/blob/moodle-error-handling/lms/services/vitalsource/factory.py#L11)


For me the tradeoff between more parameters vs more explicit dependencies is worth it.